core/src/test/scala/unit/kafka/utils/LogCaptureContext.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package unit.kafka.utils
+
+import java.util.concurrent.{CountDownLatch, TimeUnit}
+
+import org.apache.logging.log4j.Level
+import org.apache.logging.log4j.core.{LogEvent, LoggerContext}
+import org.apache.logging.log4j.test.appender.ListAppender
+
+import scala.jdk.CollectionConverters._
+
+class LogCaptureContext(listAppender: ListAppender, prevLevelMap: Map[String, Level]) extends AutoCloseable {
+
+  def setLatch(size: Int): Unit = {
+    this.listAppender.countDownLatch = new CountDownLatch(size)
+  }
+
+  @throws[InterruptedException]
+  def await(l: Long, timeUnit: TimeUnit): Unit = {
+    this.listAppender.countDownLatch.await(l, timeUnit)
+  }
+
+  def getMessages: Seq[LogEvent] = listAppender.getEvents.asScala.toSeq
+
+  override def close(): Unit = {
+    val loggerContext = LoggerContext.getContext(false)
+    loggerContext.getRootLogger.removeAppender(listAppender)
+    listAppender.stop()
+
+    // Restore previous logger levels
+    prevLevelMap.foreach { e =>
+      val loggerName = e._1
+      val level = e._2
+      loggerContext.getLogger(loggerName).setLevel(level)
+    }
+  }
+}
+
+object LogCaptureContext {
+  def apply(name: String, levelMap: Map[String, String] = Map()): LogCaptureContext = {
```

Comment:
Do we really need the caller to supply a `name`? It seems to force you to have to construct a unique name each time you want to use it, based on the test class and method name. But all that's needed is uniqueness and the removal of the appender at the end of the test, AFAICS. So just using a UUID or similar generated name would be sufficient and make call sites rather easier to read.

Reply:
> But all that's needed is uniqueness and the removal of the appender at the end of the test, AFAICS.

Exactly. However, I thought introducing randomness to the test cases is worse than some verbosity. Let's wait for others' opinions.

Reply:
I think I would also prefer generating a name in the `apply()` method. If someone uses the same name twice it might introduce an error but it's indeed easier to read without the `name` as Tom proposed.

Reply:
Got it. I will update the PR to generate the context name automatically, like `LogCaptureContext-XXXXXX`.